### PR TITLE
Reintroduced a state save menu, take two

### PR DESCRIPTION
**Pull request by npwoods**
This attempts to address the concerns that caused PR#1026 to get reverted.  Specifically, joystick button support was readded.

Take note that this (along with PR#1026) changes the menu to be driven by keyboard input rather than keycode presses.  This means that if I press 'A' on a French keyboard, I will now be saving or loading 'a.sta'.  Contrast this with the previous behavior where pressing 'A' on a French keyboard would either load or save 'q.sta' (because that keycode pats to ITEM_ID_Q).  There have been assertions that there is absolutely nothing wrong with the current behavior.  Even if we accept that, I would argue that stops being the case when you have a menu that lets you look at the directory.

Lastly, there have been assertions that PR#1026 would "partially" honor the local keyboard layout, causing a situation where the local keyboard layout would be honored on save but not load (or vice versa).  I've been unable to reproduce this behavior on my Windows machine, in either a normal or an SDL build.  If anybody has more concrete repro instructions, please advise.

### CODE REVIEW

**Comment by cuavas** [comment]
How did you actually test with French keyboard layout?  I just confirmed on a Mac that with French AZERTY layout selected, the key to the right of tab, which is the A key saves/loads q.sta - that is, it's named after the key's keycode, which is KEYCODE_Q after its label on an English QWERTY keyboard.  Likewise, with a German QWERTZ layout selected, the key to the right of left shift saves/loads y.sta - once again, the file is named for KEYCODE_Z.

I really think this it's a bad idea to change from associating files with keys to associating them with characters:

* It means the save state slots available change with the keyboard layout, e.g. while English has states `1`, `2` and `3`, French has `&`, `é` and `"` associated with those keys.
* It breaks save/load with languages that have IME edit because MAME doesn't support IME edit at all (Chinese/Japanese/Korean).
* If IME edit worked, it would require multiple keystrokes to select a slot, which would be cumbersome.  Same is true with dead keys.

If you want to make a friendly menu, you still need to work in terms of keycodes for naming the files.  If you want to provide friendly names based on the keycap in the current keyboard layout, you need to ask the OSD input module for that yourself.

Since you've repeatedly ignored these concerns which I know I've raised numerous times, I'm closing this PR again.

**Comment by npwoods** [comment]
Here is what I did (which to be fair, was on Windows and not a Mac):

I switched to an FR keyboard layout
I opened Notepad and pressed the key to the right of the Tab key, which is labelled Q on my keyboard but would be A on a French keyboard.  The letter "A" appeared
I ran MAME with my changes, and pressed SHIFT+F7 to open the state menu
I pressed the key to the right of the Tab key
I looked in my State directory.  The file 'a.sta' appeared
I pressed F7 in MAME.  The state menu appeared, and 'a.sta' was listed
I pressed the key to the right of the Tab key.  My previously saved state appeared.

I suppose this is moot, but I'm very curious why you are getting different behavior on Mac.

**Comment by cuavas** [comment]
I'm talking about current behaviour, without your current patches.  And I believe current behaviour is correct/desirable.

**Comment by npwoods** [comment]
It took me a while to understand your belief about current behavior.

I've pushed a new version that (I believe) should address your concerns.  The state filenames on disk are now "code driven", and these are now translated to local names when displayed in the UI.

File: src/frontend/mame/ui/state.cpp [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+    ui/state.cpp
+
+    Menus for saving and loading state
+
+***************************************************************************/
+
+#include "emu.h"
+#include "machine.h"
+#include "emuopts.h"
+#include "ui/state.h"
```
Comment (cuavas) [inline src/frontend/mame/ui/state.cpp]: Why explicitly `#include "machine.h"` here?  It's very much part of `emu.h`.  Also, unit header (in this case `ui/state.h`) should be right after emu.h to help avoid weird issues where headers require you to `#include` other headers before them.

File: src/frontend/mame/ui/state.cpp [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+
+
+//-------------------------------------------------
+//  keyboard_input_item_name
+//-------------------------------------------------
+
+std::string keyboard_input_item_name(input_item_id id)
+{
+	if (id >= ITEM_ID_A && id <= ITEM_ID_Z)
+		return std::string(1, (char)(id - ITEM_ID_A + 'a'));
```
Comment (cuavas) [inline src/frontend/mame/ui/state.cpp]: Can you drop the extra parens adding clutter?  This is C++ - `char(a  - b + c)` works just fine.

File: src/frontend/mame/ui/state.cpp [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+#include "machine.h"
+#include "emuopts.h"
+#include "ui/state.h"
+
+
+/***************************************************************************
+    ANONYMOUS NAMESPACE
+***************************************************************************/
+
+namespace {
```
Comment (cuavas) [inline src/frontend/mame/ui/state.cpp]: This anonymous namespace should really be inside the `namespace ui` to avoid weird dependent name lookup issues.

**Pull request merged**